### PR TITLE
Add cultivations

### PR DESCRIPTION
**Pull request by SvenVw**
## Release Notes

- **New Features**
  - Introduced a new `.env.example` file for environment variable configuration across multiple modules.
  - Added cultivation management functionalities, including options for adding and retrieving cultivations.
  - Enhanced user interface with new components for selecting cultivation and fertilizer options.

- **Improvements**
  - Updated error handling and validation in various components and functions.
  - Improved layout and user interaction in the fields and cultivation management sections.

- **Bug Fixes**
  - Resolved issues related to fetching and displaying cultivation data.

- **Documentation**
  - Updated changelogs and added example configuration files to assist users.

- **Dependencies**
  - Updated various dependencies and package versions for improved performance and security.

### CODE REVIEW

File: fdm-data/src/cultivations/catalogues/brp.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+    const catalogueSrm = brp.map(cultivation => {
+        return {
+            b_lu_source: 'brp',
+           ...cultivation
+        }
+    })
+
+    return catalogueSrm
+}
```
[the remark this answers is not in the document]
Comment (SvenVw) [reply fdm-data/src/cultivations/catalogues/brp.ts:19]: How do I add `Validate against schema if fdmSchema provides validation utilities`?

File: fdm-core/src/cultivation.d.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    b_lu: schema.cultivationsTypeSelect['b_lu'],
+    b_lu_catalogue: schema.cultivationsTypeSelect['b_lu_catalogue'],
+    b_lu_source: schema.cultivationsCatalogueTypeSelect['b_lu_source'],
+    b_lu_name: schema.cultivationsCatalogueTypeSelect['b_lu_name'],
+    b_lu_name_en: schema.cultivationsCatalogueTypeSelect['b_lu_name_en'],
+    b_lu_hcat3: schema.cultivationsCatalogueTypeSelect['b_lu_hcat3'],
+    b_lu_hcat3_name: schema.cultivationsCatalogueTypeSelect['b_lu_hcat3_name'],
+    b_sowing_date: schema.fieldSowingTypeSelectTypeSelect['b_sowing_date'],
+    b_id: schema.fieldSowingTypeSelectTypeSelect['b_id'],
+}
```
[the remark this answers is not in the document]
Comment (SvenVw) [reply fdm-core/src/cultivation.d.ts:13]: Are you sure that interfaces should be called with I starting? Various sources suggest not to do it

[the remark this answers is not in the document]
Comment (SvenVw) [reply fdm-core/src/cultivation.d.ts:13]: Partly resolved in b8e33be8b3dae706814c882abfc3bc1107a84dc5

File: fdm-data/src/cultivations/index.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+                    b_lu_catalogue: cultivation.b_lu_catalogue,
+                    b_lu_source: cultivation.b_lu_source,
+                    b_lu_name: cultivation.b_lu_name,
+                    b_lu_name_en: cultivation.b_lu_name_en,
+                    b_lu_hcat3: cultivation.b_lu_hcat3,
+                    b_lu_hcat3_name: cultivation.b_lu_hcat3_name
+                }
+            )
+        }
+    }))
```
[the remark this answers is not in the document]
Comment (SvenVw) [reply fdm-data/src/cultivations/index.ts]: Resolved in d2bcc045a24437887d5db9861708119e574d34c1

File: fdm-data/src/cultivations/index.test.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+            schema: schema
+          })
+          
+          // Run migration
+          await migrate(fdm, { migrationsFolder: migrationsFolderPath, migrationsSchema: 'fdm-migrations' })
+    })
+
+    afterAll(async () => {
+        // Clean up if necessary
+    })
```
[the remark this answers is not in the document]
Comment (SvenVw) [reply fdm-data/src/cultivations/index.test.ts:59]: The `.end` function looks to not exist on `fdm`. Are you sure?

[the remark this answers is not in the document]
Comment (SvenVw) [reply fdm-data/src/cultivations/index.test.ts:59]: Resolved in 41659042e609b19660cc0b545dfee99f2934fb88

File: fdm-core/src/cultivation.ts [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+                .delete(schema.cultivations)
+                .where(eq(schema.cultivations.b_lu, b_lu))
+
+        }
+        catch (error) {
+            tx.rollback()
+            throw new Error(`Failed to remove cultivation: ${error}`);
+        }
+    })
+}
```
[the remark this answers is not in the document]
Comment (SvenVw) [reply fdm-core/src/cultivation.ts:208]: Resolved in 3f6c7cadaeb22b196fb35f07535625302d2d242c

**Comment by SvenVw** [comment]
> **Actionable comments posted: 2**
> 🧹 Outside diff range and nitpick comments (3)
> 
> > fdm-data/src/cultivations/index.test.ts (2)> `30-38`: **Document the non-working approach**
> > > The commented code lacks explanation of why it doesn't work and when it will be fixed.
> > > ```diff
> > > - // Does not work yet :(
> > > + // TODO: Replace workaround with createFdmServer once issue #XXX is resolved
> > > + // Current blocker: <explain the specific issue>
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > 
> > > `62-84`: **Consider additional test coverage**
> > > While the happy path test is thorough, consider adding:
> > > 
> > > 1. Verification of initial database state
> > > 2. Edge cases like empty catalogue
> > > 3. Duplicate entry handling
> > > 
> > > ```diff
> > > it('should extend cultivations catalogue with brp', async () => {
> > > +   // Verify initial state
> > > +   const initialCatalogue = await getCultivationsFromCatalogue(fdm)
> > > +   expect(initialCatalogue).toHaveLength(0)
> > > 
> > >     const catalogueName = 'brp'
> > >     await extendCultivationsCatalogue(fdm, catalogueName)
> > > ```
> > 
> > 
> > 
> > fdm-core/src/cultivation.test.ts (1)> `54-146`: **Consider adding tests for edge cases**
> > > While the current test suite covers the basic CRUD operations well, consider adding tests for the following scenarios to improve coverage:
> > > 
> > > 1. Adding a duplicate cultivation
> > > 2. Adding cultivation with invalid field ID
> > > 3. Adding cultivation with invalid catalogue ID
> > > 4. Adding cultivation with invalid dates (e.g., future dates, null dates)
> > > 
> > > Would you like me to help generate these additional test cases?
> 
> 📜 Review details

Yes pleases generate the additional test cases

File: fdm-app/app/components/blocks/fields.tsx [hunk 6]
```diff
@@ -38,27 +53,82 @@ export interface fieldType {
     /** Agricultural soil type classification */
     b_soiltype_agr: string | null
     b_geojson: FeatureCollection
+    cultivationOptions: any[]
```
[the remark this answers is not in the document]
Comment (SvenVw) [reply fdm-app/app/components/blocks/fields.tsx:56]: Resolved in 3bbf38b1c5e37c24a8aad8ade30560a4a332d393

File: fdm-app/app/routes/app.addfarm.$b_id_farm.fields.tsx [hunk 7]
```diff
@@ -13,8 +13,9 @@
 
 // FDM
 import { fdm } from "../services/fdm.server";
-import { getFields, updateField } from "@svenvw/fdm-core";
+import { getCultivationsFromCatalogue, getFields, updateField } from "@svenvw/fdm-core";
 import { Button } from "@/components/ui/button";
+import { cultivationsCatalogue } from "node_modules/@svenvw/fdm-core/dist/db/schema";
```
[the remark this answers is not in the document]
Comment (SvenVw) [reply fdm-app/app/routes/app.addfarm.$b_id_farm.fields.tsx:18]: Resolved in 0b5d2e68a1053fa85da5790fa82e1c51a64a1348

File: fdm-app/app/components/blocks/fields.tsx [hunk 8]
```diff
@@ -38,27 +53,82 @@
...
+                    variant="outline"
+                    role="combobox"
+                    aria-expanded={open}
+                    className="w-[342px] justify-between"
+                >
+                    {value
+                        ? cultivationOptions.cultivationOptions.find((cultivation) => cultivation.label === value)?.label
+                        : "Selecteer hoofdgewas..."}
+                    <ChevronsUpDown className="opacity-50" />
+                </Button>
```
[the remark this answers is not in the document]
Comment (SvenVw) [reply fdm-app/app/components/blocks/fields.tsx:84]: onChange is not found

File: fdm-app/app/components/blocks/fields.tsx [hunk 9]
```diff
@@ -111,21 +181,30 @@
...
-                                            <SelectItem value="grapes">Grapes</SelectItem>
-                                            <SelectItem value="pineapple">Pineapple</SelectItem>
+                                            <SelectItem value="moerige_klei">Moerige klei</SelectItem>
+                                            <SelectItem value="rivierklei">Rivierklei</SelectItem>
+                                            <SelectItem value="dekzand">Dekzand</SelectItem>
+                                            <SelectItem value="zeeklei">Zeeklei</SelectItem>
+                                            <SelectItem value="veen">Veen</SelectItem>
+                                            <SelectItem value="loess">Löss</SelectItem>
+                                            <SelectItem value="duinzand">Duinzand</SelectItem>
+                                            <SelectItem value="maasklei">Maasklei</SelectItem>
```
[the remark this answers is not in the document]
Comment (SvenVw) [reply fdm-app/app/components/blocks/fields.tsx:202]: Resolved in 362a1689ca380a4f7dd5485c7c973a04aa592906

File: fdm-app/app/components/blocks/fields.tsx [hunk 10]
```diff
@@ -38,27 +53,82 @@ export interface fieldType {
...
+            <PopoverTrigger asChild>
+                <Button
+                    variant="outline"
+                    role="combobox"
+                    aria-expanded={open}
+                    className="w-[342px] justify-between"
+                >
+                    {value
+                        ? cultivationOptions.cultivationOptions.find((cultivation) => cultivation.label === value)?.label
+                        : "Selecteer hoofdgewas..."}
```
[the remark this answers is not in the document]
Comment (SvenVw) [reply fdm-app/app/components/blocks/fields.tsx:82]: Resolve in 3bbf38b1c5e37c24a8aad8ade30560a4a332d393

File: fdm-app/app/components/blocks/fields.tsx [hunk 11]
```diff
@@ -90,18 +160,18 @@
...
+            <Card className="w-full max-w-[750px]">
                 <Form className="space-y-6"
                     action={props.action}
                     method="post"
                     onSubmit={handleSubmit}>
                     <CardHeader>
-                        {/* <CardTitle>{props.b_name}</CardTitle>
-                        <CardDescription>{props.b_area} ha</CardDescription> */}
+                        <CardTitle>{props.b_name}</CardTitle>
+                        <CardDescription>{props.b_area} ha</CardDescription>
```
[the remark this answers is not in the document]
Comment (SvenVw) [reply fdm-app/app/components/blocks/fields.tsx]: Resolved in 460889e13911ea65024acc606f680ce5be0439e9

File: fdm-app/app/routes/app.addfarm.$b_farm_id.map.tsx [hunk 12]
```diff
@@ -123,11 +123,20 @@
...
-      const b_id = await addField(fdm, b_id_farm, b_id_name, b_id_source, b_geometry, null, null, null)
 
-      return b_id
+      const b_id = await addField(fdm, b_id_farm, b_id_name, b_id_source, b_geometry, b_manage_start, null, null)
+      const b_lu = await addCultivation(fdm, b_lu_catalogue, b_id, b_date_sowing)
+
+      return {
+        b_id: b_id,
+        b_lu: b_lu
+      }
```
[the remark this answers is not in the document]
Comment (SvenVw) [reply fdm-app/app/routes/app.addfarm.$b_farm_id.map.tsx]: Resolved in 80db2e705d15192b7fe90f983416f2c644985474

File: fdm-core/src/cultivation.test.ts [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+            const b_lu = await addCultivation(fdm, b_lu_catalogue, b_id, b_sowing_date)
+
+
+            await removeCultivation(fdm, b_lu)
+
+            await expect(getCultivation(fdm, b_lu)).rejects.toThrowError('Cultivation does not exist')
+        })
+
+
+    })
```
[the remark this answers is not in the document]
Comment (SvenVw) [reply fdm-core/src/cultivation.test.ts:146]: Yes please generate the additional unit tests

File: fdm-core/src/cultivation.ts [hunk 14]
```diff
@@ -0,0 +1,283 @@
...
+        b_lu_hcat3: schema.cultivationsCatalogueTypeInsert['b_lu_hcat3'],
+        b_lu_hcat3_name: schema.cultivationsCatalogueTypeInsert['b_lu_hcat3_name']
+    }
+): Promise<void> {
+
+    // Insert the cultivation in the db
+    await fdm
+        .insert(schema.cultivationsCatalogue)
+        .values(properties)
+}
```
[the remark this answers is not in the document]
Comment (SvenVw) [reply fdm-core/src/cultivation.ts:49]: Resolved in d6b8294ffcaa542d1af1482458cc18d625e570b2

File: fdm-app/app/components/blocks/fields.tsx [hunk 15]
```diff
@@ -38,27 +52,85 @@ export interface fieldType {
...
+                                    />
+                                </CommandItem>
+                            ))}
+                        </CommandGroup>
+                    </CommandList>
+                </Command>
+            </PopoverContent>
+        </Popover>
+    )
+}
```
[the remark this answers is not in the document]
Comment (SvenVw) [reply fdm-app/app/components/blocks/fields.tsx]: Please check again with updated code

File: fdm-app/eslint.config.js [hunk 16]
```diff
@@ -0,0 +1,10 @@
+import globals from 'globals'
+import pluginJs from '@eslint/js'
+import tseslint from 'typescript-eslint'
+
+export default [
+  { files: ['src/**/*.{js,mjs,cjs,ts}'] },
```
[the remark this answers is not in the document]
Comment (SvenVw) [reply fdm-app/eslint.config.js:6]: Resolved in e1aadf58f9718d6a3c7681a1ad800a093849cb2c
